src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs

```diff
@@ -211,36 +175,51 @@ public int TakeData(byte[] buffer, int dataOffset, int size)
                 size = _length - _offset;
             }
 
-            Buffer.BlockCopy(_data, _offset, buffer, dataOffset, size);
+            Buffer.BlockCopy(_data, _header + _offset, buffer, dataOffset, size);
             _offset += size;
             return size;
         }
 
+
+        /// <summary>
+        /// Set the MARS SMUX header information for this packet
+        /// </summary>
+        /// <param name="header">the header to write into the packet</param>
+        public void SetHeader(SNISMUXHeader header)
```

Comment:
Ideally I would like to have SNIPacket agnostic of MARS. It should be a container for the bytes and have the information about the stream that it should write to. 
Can you refactor this, so that SNIPacket is agnostic?

Reply:
I've split this into two functions `GetHeadrBuffer` and `SetHeaderActive` because the only way to do it with a single function i could see uses a delegate and i don't want to pay the allocation cost.  It's used from the mars handle now.

To remove the knowledge of smux from other places i've changed the SNIHandle so that it has a reserved header property which indicates how large a header it wants, all standard handles return 0 and the marsh handle returns the smux header length so the handle now determines the header reservation and no purpose is specified. These changes flow through the other files.

Reply:
Thanks, this is much nicer approach.